### PR TITLE
Add support for FSDP+QLoRA and DeepSpeed ZeRO3+QLoRA

### PR DESCRIPTION
**Pull request by pacman100**
### What does this PR do? 
1. `prepare_model_for_kbit_training` and `peft_module_casting_to_bf16` should be disabled when using FSDP+QLoRA or DeepSpeed ZeRO3+QLoRA.

This PR should be merged after Transformers PR https://github.com/huggingface/transformers/pull/29587

### CODE REVIEW

**Comment by pacman100** [comment]
cc @younesbelkada

File: trl/trainer/sft_trainer.py [hunk 1]
```diff
@@ -186,7 +186,15 @@ def __init__(
                     inspect.signature(prepare_model_for_kbit_training).parameters
                 )
                 gradient_checkpointing_kwargs = getattr(args, "gradient_checkpointing_kwargs", None) or {}
-                if getattr(model, "is_loaded_in_8bit", False) or getattr(model, "is_loaded_in_4bit", False):
+                is_sharded_qlora = False
+                if getattr(model, "is_loaded_in_4bit", False):
```
Comment (younesbelkada) [inline trl/trainer/sft_trainer.py:190]:
```suggestion
                # Below is to support QLoRA + FSDP / DS-Zero3 - one should never call 
                # peft_module_casting_to_bf16 or prepare_model_for_kbit_training when doing 
                # QLoRA + FSDP / DS-Zero3
                if getattr(model, "is_loaded_in_4bit", False):
```

**Review by younesbelkada — APPROVED** [review]
Thanks so much !

**Comment by younesbelkada** [comment]
I can't repro the CI failure locally and on main, will fix in a follow up PR!

**Pull request merged**